### PR TITLE
Add feature to detect honeypot and version

### PR DESCRIPTION
**Pull request by ZupaPomidorowa**
# Final report: GSoC 2024

This PR has been created as part of the Google Summer of Code in the "Honeyscanner - A vulnerability analyzer for Honeypots" project. The goal of my project is to automatically detect honeypot and it's version.

**Usage**
`python3 main.py --target_ip 127.0.0.1 --detect  `

**Example response**
`This host is probably Conpot honeypot`
`Conpot version Release_0.6.0`

### CODE REVIEW

File: honeyscanner/detect_honeypot/detect_honeypot.py [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+    port_cowrie_kippo = 2222
+    dionaea_ports = [21, 23, 42, 53, 80, 135, 443, 445, 1433, 1723, 1883, 3306, 5060, 9100, 11211, 27017]
+    conpot_ports = [2121, 5020, 10201, 44818]
+
+    if port_cowrie_kippo in open_ports:
+        resp = port_2222(ip)
+        if resp == 'cowrie':
+            check_cowrie(ip)
+            print('This host is probably Cowrie honeypot')
+            version = get_latest_version('cowrie')
```
Comment (aristofanischionis) [inline honeyscanner/detect_honeypot/detect_honeypot.py:263]: It is hard to identify the honeypot version that runs on an IP address during a black-box assessment, but accepting that the deployed version is the latest, is a `wild guess`. Honeyscanner needs to identify as accurately as possible older versions of open-source honeypots, as they are more likely to be running on outdated and vulnerable libraries.

File: honeyscanner/Pipfile [hunk 2]
```diff
@@ -12,6 +12,8 @@ packaging = "*"
 paramiko = "*"
 pygithub = "*"
 python3-nmap = "*"
+socket = "*"
+ssl = "*"
```
Comment (aristofanischionis) [inline honeyscanner/Pipfile:16]: Why do we need the `ssl` library?

File: honeyscanner/detect_honeypot/detect_honeypot.py [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+            print(f'Cowrie version {version}')
+        elif resp == 'kippo':
+            print('This host is probably Kippo honeypot')
+            version = get_latest_version('kippo')
+            print(f'Kippo version {version}')
+    elif set(dionaea_ports) <= set(open_ports):
+        print('The same open ports as in Dionaea honeypot')
+        if check_dionaea(ip):
+            print('This host is probably Dionaea honeypot')
```
Comment (aristofanischionis) [inline honeyscanner/detect_honeypot/detect_honeypot.py:272]: Add one more printing describing the reason, that you tried connecting to 3 ports and was successful.

**Review by aristofanischionis — CHANGES_REQUESTED** [review]

File: honeyscanner/detect_honeypot/detect_honeypot.py (hunk 1 above)
Comment (ZupaPomidorowa) [reply honeyscanner/detect_honeypot/detect_honeypot.py:263]: Do you have any ideas how to do this?

File: honeyscanner/Pipfile (hunk 2 above)
Comment (ZupaPomidorowa) [reply honeyscanner/Pipfile:16]: Because on port 455 in Dionaea honeypot is running http server and I want to get ssl cert which contains 'dionaea'

Comment (Swangeon) [reply honeyscanner/Pipfile:16]: The `socket = "*"` line can be removed as socket is a built-in module.

**Pull request closed without merging**